src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -992,38 +1059,50 @@
     :as fact-data} :- facts-schema]
   (jdbc/with-db-transaction []
     (let [factset-id (certname-to-factset-id certname)
-          paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
-          pathstrs (map #(facts/factpath-to-string (first %))
-                        paths-and-valuemaps)
-          pathstr->valuemap (into {} (map (fn [pathstr [_ vm]] [pathstr vm])
-                                          pathstrs
-                                          paths-and-valuemaps))
+          paths-and-values (facts/facts->paths-and-values values)
+          pathstrs (map (comp facts/factpath-to-string first) paths-and-values)
+          pathstr->value (into {} (map (fn [pathstr [_ v]] [pathstr v])
+                                       pathstrs
+                                       paths-and-values))
+          missing (Object.)
           existing-status ;; :changed :unchanged :gone
           (jdbc/call-with-query-rows
            [(str "select"
                  "    path, id, value_type_id, large_value_hash,"
-                 "    case when large_value_hash is null then value end as value"
+                 "    case when large_value_hash is null "
+                 "              and value_type_id = " db-fv-struct
+                 "           then value"
+                 "    end as value,"
+                 "    case when large_value_hash is null then value_string"
+                 "    end as value_string,"
+                 "    value_integer, value_float, value_boolean"
                  "  from facts as f"
                  "  inner join fact_paths fp on f.fact_path_id = fp.id"
                  "  where f.factset_id = ?")
             factset-id]
-           (fn [rows]
+           {:as-arrays? true}
+           (fn [[col-names & rows]]
              (doall
-              (for [{:keys [path id] :as row} rows]
-                (if-let [new-vm (pathstr->valuemap path)]
-                  (if (fact-value-changed? row new-vm)
-                    [:changed path id new-vm]
-                    [:unchanged path id])
-                  [:gone path id])))))
-
+              (for [[path id type-id hash v vstr vint vfloat vbool :as row] rows]
+                ;; Classify row as :changed :unchanged: :gone
+                (let [new-val (pathstr->value path missing)]
```

Comment:
I find missing here non-idiomatic. I'd also pretty use get if I'm using a default, so I think typically you'd see
`(get pathstr->value path ::not-found)` or something similar.